[FSDP] Cache _full_param_padded

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66958

It looks like this is always created in the same way during lazy init
for every iteration. But it is pretty much the same tensor every iteration, so
we can cache it.

Differential Revision: [D31783780](https://our.internmc.facebook.com/intern/diff/D31783780/)

rohan-varma committed:
[FSDP] Cache _full_param_padded

It looks like this is always created in the same way during lazy init
for every iteration. But it is pretty much the same tensor every iteration, so
we can cache it.

Differential Revision: [D31783780](https://our.internmc.facebook.com/intern/diff/D31783780/)

[ghstack-poisoned]

zhaojuanmao reviewed (commented):
looks good, thanks for this optimization! This may improve memory fragmentation and potentially improve cudaMalloc performance when memory usage is high. 

would you please run QPS and accuracy benchmarks as well?

rohan-varma committed:
Update on "[FSDP] Cache _full_param_padded"


It looks like this is always created in the same way during lazy init
for every iteration. But it is pretty much the same tensor every iteration, so
we can cache it.

Differential Revision: [D31783780](https://our.internmc.facebook.com/intern/diff/D31783780/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Cache _full_param_padded"


It looks like this is always created in the same way during lazy init
for every iteration. But it is pretty much the same tensor every iteration, so
we can cache it.

Differential Revision: [D31783780](https://our.internmc.facebook.com/intern/diff/D31783780/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Cache _full_param_padded"


It looks like this is always created in the same way during lazy init
for every iteration. But it is pretty much the same tensor every iteration, so
we can cache it.

Differential Revision: [D31783780](https://our.internmc.facebook.com/intern/diff/D31783780/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Cache _full_param_padded"


It looks like this is always created in the same way during lazy init
for every iteration. But it is pretty much the same tensor every iteration, so
we can cache it.

Differential Revision: [D31783780](https://our.internmc.facebook.com/intern/diff/D31783780/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Cache _full_param_padded"


It looks like this is always created in the same way during lazy init
for every iteration. But it is pretty much the same tensor every iteration, so
we can cache it.

Differential Revision: [D31783780](https://our.internmc.facebook.com/intern/diff/D31783780/)

[ghstack-poisoned]

rohan-varma commented:
This is actually not useful, as `_init_param_attributes` returns early if the param is already sharded, which it always is after the first iteration. So this caching doesn't actually result in any saving, closing this out.
